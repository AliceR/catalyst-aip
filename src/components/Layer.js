import { useEffect } from 'react'
import PropTypes from 'prop-types'

export default function Layer({ id, spec, map }) {
  useEffect(() => {
    const layer = map.addLayer(spec)

    return () => {
      if (layer) map.removeLayer(id)
    }
  }, [])

  return null
}

Layer.propTypes = {
  id: PropTypes.string.isRequired,
  map: PropTypes.object,
}
