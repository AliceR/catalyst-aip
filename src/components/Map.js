import React, { useLayoutEffect, useState, useRef } from 'react'
import PropTypes from 'prop-types'
import mapbox from 'mapbox-gl'
import styled from 'styled-components'

const MapContainer = styled.div`
  height: 100%;
  flex-grow: 1;
`

export default function Map({ children }) {
  const ref = useRef()
  const [map, setMap] = useState(null)

  useLayoutEffect(() => {
    const m = new mapbox.Map({
      container: ref.current,
      style: 'mapbox://styles/mapbox/streets-v11',
      zoom: 7,
      center: [37.85335, 0.44014],
    })

    m.on('load', () => {
      setMap(m)

      if (process.env.NODE_ENV === 'development') {
        // makes map accessible in console for debugging
        window.map = m
      }
    })

    return () => {
      if (map) {
        map.remove()
      }
    }
  }, [])

  return (
    <MapContainer id='map' ref={ref}>
      {map &&
        children &&
        React.Children.map(children, (child) =>
          React.cloneElement(child, {
            map,
          })
        )}
    </MapContainer>
  )
}

Map.propTypes = {
  children: PropTypes.node,
}
