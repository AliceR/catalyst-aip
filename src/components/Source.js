import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'

export default function Source({ id, type, tilesetid, map, children }) {
  const [source, setSource] = useState(null)

  const tileSize = type === 'raster' ? 256 : 512

  useEffect(() => {
    let s = map.getSource(`${id}-source`)
    if (!s) {
      s = map.addSource(id, {
        type,
        url: `mapbox://${tilesetid}`,
        tileSize,
      })
    }

    setSource(s)

    return () => {
      if (source) map.removeSource(id)
    }
  }, [])
  return (
    <>
      {source &&
        children &&
        React.Children.map(children, (child) =>
          React.cloneElement(child, {
            map: map,
            sourceId: id,
          })
        )}
    </>
  )
}

Source.propTypes = {
  id: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['raster', 'vector', 'geojson']),
  tilesetid: PropTypes.string.isRequired,
  map: PropTypes.object,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
}
