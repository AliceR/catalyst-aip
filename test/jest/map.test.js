import React from 'react'
import renderer from 'react-test-renderer'
import mapbox from 'mapbox-gl'

import Map from '../../src/components/Map'
import Source from '../../src/components/Source'
import Layer from '../../src/components/Layer'

let container
let mapWrapper, sourceWrapper, layerWrapper

const source = {
  id: 'test-source',
  layer: 'test-source',
  tilesetid: 'test.12345',
}

const layer = {
  id: 'test-layer',
  source: 'test-source',
  'source-layer': 'test-source',
  type: 'vector',
}

describe('Map', () => {
  beforeEach(() => {
    renderer.act(() => {
      mapWrapper = renderer.create(<Map style={{ height: 200 }} />, {
        createNodeMock: (element) => {
          container = element
          return element
        },
      })
    })
  })

  it('renders a styled container', () => {
    expect(mapWrapper.toJSON()).toMatchSnapshot()
  })

  it('adds a map', () => {
    expect(mapbox.Map).toHaveBeenCalledWith({
      container,
      style: `mapbox://styles/mapbox/streets-v11`,
      zoom: 7,
      center: [37.85335, 0.44014],
    })
  })
})

describe('Source', () => {
  it('adds a map source', () => {
    renderer.act(() => {
      sourceWrapper = renderer.create(
        <Source
          id={source.id}
          type='raster'
          tilesetid={source.tilesetid}
          map={mapbox.Map()}
        />
      )
    })

    expect(sourceWrapper.toTree().props.map.addSource).toHaveBeenCalledWith(
      source.id,
      {
        type: 'raster',
        url: 'mapbox://test.12345',
        tileSize: 256,
      }
    )
  })
})

describe('Layer', () => {
  beforeEach(() => {
    renderer.act(() => {
      layerWrapper = renderer.create(
        <Layer
          id={layer.id}
          spec={layer}
          sourceId={source.id}
          map={mapbox.Map()}
        />
      )
    })
  })
  it('adds a map layer', () => {
    expect(layerWrapper.toTree().props.map.addLayer).toHaveBeenCalledWith({
      id: layer.id,
      source: source.id,
      'source-layer': source.layer,
      type: 'vector',
    })
  })
})
