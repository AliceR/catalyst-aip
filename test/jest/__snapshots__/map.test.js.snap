// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders a styled container 1`] = `
.c0 {
  grid-column: 4 / span 9;
}

<section
  className="c0"
  id="map"
/>
`;
